### ISSUE TITLE
False positives testing for array emptiness

### ISSUE THREAD

**Issue by flaviovs** [vimeo/psalm#10825]
https://psalm.dev/r/4609d92f6f

Note: using `empty()` because it's a common pattern in PHP apps to assert nullable arrays.  However, removing `empty()` (i.e. doing `if ($this->arr)...`) **still generates warnings**, albeit different ones:

```
ERROR: [TypeDoesNotContainType](https://psalm.dev/056) - 13:13 - Operand of type array<never, never> is always falsy
ERROR: [TypeDoesNotContainType](https://psalm.dev/056) - 13:13 - Type array<never, never> for $this->arr is always !falsy
```

**Comment by flaviovs** [comment]
> Could you provide a full example? When I call something that modifies it as your comment says, it works fine: https://psalm.dev/r/d12fcff59a

Hi @kkmuffme - I'm not sure I follow.

The snippet I posted is a minimal example that reproduces the issue. It is from a big app codebase that started to emit warnings after upgrading from 5.19.0 to latest Psalm. I don't see anything wrong with that code.

IMHO your changed code not issuing any warnings just confirm the problem, since your version do the same thing, i.e. `$arr === [] || $a === null` is almost the same as `empty($arr)`, just more verbose.

**Comment by flaviovs** [comment]
Here's another minimal example that reproduces the issue, extracted from the same app that started to emit warnings after upgrade:

https://psalm.dev/r/44845ca739

**Comment by ArtemGoutsoul** [comment]
Maybe a somewhat more minimal case:

```php
/**
 * @param true[] $array
 */
function check_key_in_array(array $array, int $key): void
{
    if (!empty($array[$key])) echo "key $key not empty";
}
```
gives: `ERROR: [RedundantConditionGivenDocblockType](https://psalm.dev/156) - 8:9 - Operand of type true is always truthy`

the element is ALWAYS truthy when present, but it can be missing -> empty -> falsy, since the keys are not specified.


https://psalm.dev/r/557faa3f96

**Comment by ArtemGoutsoul** [comment]
This regression seems to be related to:
- https://github.com/vimeo/psalm/issues/10578
- https://github.com/vimeo/psalm/issues/10715
- https://github.com/vimeo/psalm/issues/10716

It's still reproducible in 5.23.1

**Comment by ArtemGoutsoul** [comment]
@kkmuffme you are right, maybe it's a separate issue, but should not be https://github.com/vimeo/psalm/issues/10715, since this issue is still present in 5.23.1
 
If the fix for https://github.com/vimeo/psalm/issues/10715 is in the upcoming release and not in 5.23.1, then my comments are a brainfart on my part and should be erased, I should just wait for the next release.

**Comment by M1ke** [comment]
My team are debating this assertion

> `$arr === [] || $arr === null` is almost the same as `empty($arr)`

Does anyone have a value of `$arr` (where arr meets type definition `array|null`) where the `empty` case returns a different bool to the `||` case?

Tried a set here: https://3v4l.org/VlLK6#v8.3.8

**Comment by ArtemGoutsoul** [comment]
> My team are debating this assertion
> 
> > `$arr === [] || $arr === null` is almost the same as `empty($arr)`
> 
> Does anyone have a value of `$arr` (where arr meets type definition `array|null`) where the `empty` case returns a different bool to the `||` case?
> 
> Tried a set here: https://3v4l.org/VlLK6#v8.3.8

fFor a defined $arr that is `array|null` the expression `$arr === [] || $arr === null` is identical to `empty($arr)`, since empty essentially casts to bool (and returns false for undefined vars):
- https://www.php.net/manual/en/function.empty.php
- https://www.php.net/manual/en/language.types.boolean.php#language.types.boolean.casting

**Comment by M1ke** [comment]
OK great; I was wondering about the "almost the same" comment because I really couldn't see why they were not identical. Someone pointed out that empty is safe for checking undefined values, but that's a different condition.

**Comment by flaviovs** [comment]
OP here. I've done some more tests an here's what I came up:

Leaving `null` check out, same results: https://psalm.dev/r/ba93e8414a

Using `count($arr) === 0` or `$arr === []`, same results:

* https://psalm.dev/r/a93bf069be
* https://psalm.dev/r/ee7c8a7f07

So the issue may not be even related to `empty()` after all.
